File: tests/unit/test_types.py
"""Tests for src.tyche.types module."""
from unittest.mock import patch

from src.tyche.types import (
    ADMIN_PORT_DEFAULT,
    HEARTBEAT_INTERVAL,
    HEARTBEAT_LIVENESS,
    BackpressureStrategy,
    DurabilityLevel,
    Endpoint,
    EventType,
    Interface,
    InterfacePattern,
    MessageType,
    ModuleId,
    ModuleInfo,
)


class TestModuleId:
    def test_generate_format(self):
        """Module ID has format family_6hexchars."""
        mid = ModuleId.generate("testfamily")
        parts = mid.split("_")
        assert len(parts) == 2
        assert parts[0] == "testfamily"
        assert len(parts[1]) == 6
        assert all(c in "0123456789abcdef" for c in parts[1])

    def test_generate_default_family(self):
        """Default family is 'unknown'."""
        mid = ModuleId.generate()
        assert mid.startswith("unknown_")

    def test_generate_unique(self):
        """Multiple calls produce different IDs."""
        ids = {ModuleId.generate("fam") for _ in range(100)}
        assert len(ids) == 100

    @patch("src.tyche.types.secrets.token_hex")
    def test_generate_uses_token_hex(self, mock_token_hex):
        """Verify token_hex is called with 3 for 6 hex chars."""
        mock_token_hex.return_value = "aabbcc"
        mid = ModuleId.generate("fam")
        mock_token_hex.assert_called_once_with(3)
        assert mid == "fam_aabbcc"


class TestEventType:
    def test_event_type_values(self):
        assert EventType.REQUEST.value == "request"
        assert EventType.RESPONSE.value == "response"
        assert EventType.EVENT.value == "event"
        assert EventType.HEARTBEAT.value == "heartbeat"
        assert EventType.REGISTER.value == "register"
        assert EventType.ACK.value == "ack"


class TestInterfacePattern:
    def test_pattern_values(self):
        assert InterfacePattern.ON.value == "on"
        assert InterfacePattern.SEND.value == "send"
        assert InterfacePattern.HANDLE.value == "handle"
        assert InterfacePattern.REQUEST.value == "request"


class TestBackpressureStrategy:
    def test_strategy_values(self):
        assert BackpressureStrategy.DROP_OLDEST.value == "drop_oldest"
        assert BackpressureStrategy.DROP_NEWEST.value == "drop_newest"
        assert BackpressureStrategy.BLOCK_PRODUCER.value == "block_producer"


class TestDurabilityLevel:
    def test_level_values(self):
        assert DurabilityLevel.BEST_EFFORT.value == 0
        assert DurabilityLevel.ASYNC_FLUSH.value == 1
        assert DurabilityLevel.SYNC_FLUSH.value == 2


class TestMessageType:
    def test_message_type_values(self):
        assert MessageType.COMMAND.value == "cmd"
        assert MessageType.EVENT.value == "evt"
        assert MessageType.HEARTBEAT.value == "hbt"
        assert MessageType.REGISTER.value == "reg"
        assert MessageType.ACK.value == "ack"
        assert MessageType.RESPONSE.value == "resp"
        assert MessageType.REQUEST.value == "req"


class TestEndpoint:
    def test_str_format(self):
        ep = Endpoint("127.0.0.1", 5555)
        assert str(ep) == "tcp://127.0.0.1:5555"

    def test_str_different_port(self):
        ep = Endpoint("0.0.0.0", 8080)
        assert str(ep) == "tcp://0.0.0.0:8080"


class TestInterface:
    def test_defaults(self):
        iface = Interface(name="on_test", pattern=InterfacePattern.ON, event_type="test")
        assert iface.durability == DurabilityLevel.ASYNC_FLUSH
        assert iface.backpressure == BackpressureStrategy.DROP_OLDEST
        assert iface.max_queue_depth == 10000
        assert iface.wait_timeout is None

    def test_custom_values(self):
        iface = Interface(
            name="handle_compute",
            pattern=InterfacePattern.HANDLE,
            event_type="compute",
            durability=DurabilityLevel.SYNC_FLUSH,
            backpressure=BackpressureStrategy.DROP_NEWEST,
            max_queue_depth=500,
            wait_timeout=10.0,
        )
        assert iface.durability == DurabilityLevel.SYNC_FLUSH
        assert iface.backpressure == BackpressureStrategy.DROP_NEWEST
        assert iface.max_queue_depth == 500
        assert iface.wait_timeout == 10.0


class TestModuleInfo:
    def test_defaults(self):
        info = ModuleInfo(module_id="mod_1", interfaces=[], metadata={})
        assert info.family_name == ""
        assert info.admin_handlers == {}

    def test_with_admin_handlers(self):
        def handler():
            return {}

        info = ModuleInfo(
            module_id="mod_1",
            interfaces=[],
            metadata={"key": "value"},
            family_name="test_family",
            admin_handlers={"health": handler},
        )
        assert info.family_name == "test_family"
        assert info.admin_handlers["health"] == handler


class TestConstants:
    def test_heartbeat_interval(self):
        assert HEARTBEAT_INTERVAL == 1.0

    def test_heartbeat_liveness(self):
        assert HEARTBEAT_LIVENESS == 3

    def test_admin_port_default(self):
        # Admin port is base_port + 3 (registration=5555 → admin=5558),
        # matching the C++ engine port layout documented in
        # src/tyche/cpp/engine/main.cpp.
        assert ADMIN_PORT_DEFAULT == 5558

File: tests/cpp/test_types.cpp
// Unit tests for tyche types: ModuleId, enums, Endpoint, string conversions.

#include <gtest/gtest.h>

#include <regex>
#include <set>
#include <string>

#include "tyche/cpp/types.h"

namespace tyche {
namespace {

// ── ModuleId Generation ───────────────────────────────────────────────

TEST(ModuleIdTest, GenerateFormat) {
    std::string id = ModuleId::generate("gateway");

    // Should match: gateway_[6 hex chars]
    std::regex pattern("gateway_[0-9a-f]{6}");
    EXPECT_TRUE(std::regex_match(id, pattern)) << "Got: " << id;
}

TEST(ModuleIdTest, GenerateWithDifferentFamilies) {
    std::string id1 = ModuleId::generate("greeks");
    std::string id2 = ModuleId::generate("static_data");

    EXPECT_TRUE(id1.find("greeks_") == 0);
    EXPECT_TRUE(id2.find("static_data_") == 0);
}

TEST(ModuleIdTest, GenerateDefaultFamily) {
    std::string id = ModuleId::generate();
    EXPECT_TRUE(id.find("unknown_") == 0);
}

TEST(ModuleIdTest, GenerateUniqueness) {
    std::set<std::string> ids;
    for (int i = 0; i < 100; ++i) {
        ids.insert(ModuleId::generate("test"));
    }
    // With 24 bits of randomness, 100 IDs should be unique
    EXPECT_EQ(ids.size(), 100u);
}

// ── MessageType String Conversion ─────────────────────────────────────

TEST(TypesTest, MessageTypeToStr) {
    EXPECT_STREQ(message_type_to_str(MessageType::COMMAND), "cmd");
    EXPECT_STREQ(message_type_to_str(MessageType::EVENT), "evt");
    EXPECT_STREQ(message_type_to_str(MessageType::HEARTBEAT), "hbt");
    EXPECT_STREQ(message_type_to_str(MessageType::REGISTER), "reg");
    EXPECT_STREQ(message_type_to_str(MessageType::ACK), "ack");
    EXPECT_STREQ(message_type_to_str(MessageType::RESPONSE), "resp");
    EXPECT_STREQ(message_type_to_str(MessageType::REQUEST), "req");
}

TEST(TypesTest, MessageTypeFromStr) {
    EXPECT_EQ(message_type_from_str("cmd"), MessageType::COMMAND);
    EXPECT_EQ(message_type_from_str("evt"), MessageType::EVENT);
    EXPECT_EQ(message_type_from_str("hbt"), MessageType::HEARTBEAT);
    EXPECT_EQ(message_type_from_str("reg"), MessageType::REGISTER);
    EXPECT_EQ(message_type_from_str("ack"), MessageType::ACK);
    EXPECT_EQ(message_type_from_str("resp"), MessageType::RESPONSE);
    EXPECT_EQ(message_type_from_str("req"), MessageType::REQUEST);
}

TEST(TypesTest, MessageTypeFromStrUnknownDefaultsToEvent) {
    EXPECT_EQ(message_type_from_str("unknown"), MessageType::EVENT);
    EXPECT_EQ(message_type_from_str(""), MessageType::EVENT);
}

// ── InterfacePattern String Conversion ────────────────────────────────

TEST(TypesTest, InterfacePatternToStr) {
    EXPECT_STREQ(interface_pattern_to_str(InterfacePattern::ON), "on");
    EXPECT_STREQ(interface_pattern_to_str(InterfacePattern::SEND), "send");
    EXPECT_STREQ(interface_pattern_to_str(InterfacePattern::HANDLE), "handle");
    EXPECT_STREQ(interface_pattern_to_str(InterfacePattern::REQUEST), "request");
}

TEST(TypesTest, InterfacePatternFromStr) {
    EXPECT_EQ(interface_pattern_from_str("on"), InterfacePattern::ON);
    EXPECT_EQ(interface_pattern_from_str("send"), InterfacePattern::SEND);
    EXPECT_EQ(interface_pattern_from_str("handle"), InterfacePattern::HANDLE);
    EXPECT_EQ(interface_pattern_from_str("request"), InterfacePattern::REQUEST);
}

TEST(TypesTest, InterfacePatternFromStrUnknownDefaultsToOn) {
    EXPECT_EQ(interface_pattern_from_str("foo"), InterfacePattern::ON);
    EXPECT_EQ(interface_pattern_from_str(""), InterfacePattern::ON);
}

// ── Endpoint ──────────────────────────────────────────────────────────

TEST(EndpointTest, DefaultConstruction) {
    Endpoint ep;
    EXPECT_EQ(ep.host, "");
    EXPECT_EQ(ep.port, 0);
}

TEST(EndpointTest, ParameterizedConstruction) {
    Endpoint ep("127.0.0.1", 5555);
    EXPECT_EQ(ep.host, "127.0.0.1");
    EXPECT_EQ(ep.port, 5555);
}

TEST(EndpointTest, ToString) {
    Endpoint ep("127.0.0.1", 5555);
    EXPECT_EQ(ep.to_string(), "tcp://127.0.0.1:5555");

    Endpoint ep2("0.0.0.0", 8080);
    EXPECT_EQ(ep2.to_string(), "tcp://0.0.0.0:8080");
}

// ── Interface Struct ──────────────────────────────────────────────────

TEST(InterfaceTest, DefaultValues) {
    Interface iface;
    EXPECT_EQ(iface.pattern, InterfacePattern::ON);
    EXPECT_EQ(iface.durability, DurabilityLevel::ASYNC_FLUSH);
    EXPECT_EQ(iface.backpressure, BackpressureStrategy::DROP_OLDEST);
    EXPECT_EQ(iface.max_queue_depth, 10000);
}

// ── Constants ─────────────────────────────────────────────────────────

TEST(TypesTest, HeartbeatConstants) {
    EXPECT_DOUBLE_EQ(HEARTBEAT_INTERVAL, 1.0);
    EXPECT_EQ(HEARTBEAT_LIVENESS, 3);
    // Admin port is base_port + 3 (registration=5555 → admin=5558),
    // matching the port layout documented in src/tyche/cpp/engine/main.cpp.
    EXPECT_EQ(ADMIN_PORT_DEFAULT, 5558);
}

// ── DurabilityLevel Values ────────────────────────────────────────────

TEST(TypesTest, DurabilityLevelValues) {
    EXPECT_EQ(static_cast<int>(DurabilityLevel::BEST_EFFORT), 0);
    EXPECT_EQ(static_cast<int>(DurabilityLevel::ASYNC_FLUSH), 1);
    EXPECT_EQ(static_cast<int>(DurabilityLevel::SYNC_FLUSH), 2);
}

}  // namespace
}  // namespace tyche
